---
title: "Differential_Expression"
author: "Srilakshmi"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Differential Expression Analysis between Controls and Patients
## Merge Healthy OSN and Disease
```{r}
disease.subset_OSN <- readRDS("../outputs/RDS_files/disease/disease.subset_OSN.rds")
samples.seurat <- readRDS('../outputs/RDS_files/samples/samples.seuratpseudorank_june.rds')
```

```{r}
disease.subset_OSN@assays$SCT <- NULL
DefaultAssay(disease.subset_OSN) <- "RNA"
samples.seurat@assays$SCT <- NULL
DefaultAssay(samples.seurat) <- "RNA"

disam.seuratmerge <- merge(x = disease.subset_OSN, y = samples.seurat, merge.data = TRUE,
  project = "disease_sample")
```

## Run SCTransform 
```{r}
disam.seuratmerge <- SCTransform(disam.seuratmerge, method = "glmGamPoi", 
                vars.to.regress = c("S.Score","G2M.Score"), verbose = FALSE)
```

## Run PCA and UMAP
```{r}
## Perform UMAP reduction
disam.seuratmerge <- RunPCA(disam.seuratmerge)
disam.seuratmerge <- RunUMAP(disam.seuratmerge, dims = 1:50)
```

## Integration to reduce batch effects
```{r}
# Harmony method
disam.seuratmerge <- 
  IntegrateLayers(object = disam.seuratmerge, method = HarmonyIntegration, orig.reduction = "pca", new.reduction = "harmony", verbose = FALSE
)

#options(future.globals.maxSize = 6 * 1024^3)

# RPCA method
#disam.seuratmerge <- 
# IntegrateLayers(object = disam.seuratmerge, method = RPCAIntegration, orig.reduction = #"pca", new.reduction = "rpca", verbose = TRUE, normalization.method = "SCT"
#)

#CCA method 
#disam.seuratmerge <- 
#  IntegrateLayers(object = disam.seuratmerge, method = CCAIntegration, orig.reduction = "pca", #new.reduction = "cca", verbose = TRUE, normalization.method = "SCT"
#)
```

## Run UMAP
```{r}
disam.seuratmerge <- RunUMAP(disam.seuratmerge, dims = 1:50, reduction = "harmony", reduction.name = "harmony_umap")
```

## Cluster
```{r}
res.range <- seq(0.4, 2.4, 0.2)

DefaultAssay(disam.seuratmerge) <- "SCT"
disam.seuratmerge <- FindNeighbors(disam.seuratmerge, dims = 1:50)

for (res in res.range) {
    disam.seuratmerge <- FindClusters(disam.seuratmerge, resolution = res)
    U <- DimPlot(disam.seuratmerge, reduction = "harmony_umap", label = TRUE) +
      ggtitle(paste("Resolution:", res))
    print(U)
}
```

```{r}
clustree::clustree(disam.seuratmerge, prefix = "SCT_snn_res.")
```

```{r}
res <- 1.2  #change accordingly
disam.seuratmerge <- FindNeighbors(disam.seuratmerge, verbose = F)
disam.seuratmerge <- FindClusters(disam.seuratmerge, resolution = res, verbose = F)
```

```{r}
DimPlot(disam.seuratmerge, reduction = 'harmony_umap', group.by = 'label')
DimPlot(disam.seuratmerge, reduction = 'harmony_umap', group.by = 'orig.ident')
```

##Add condition variables
```{r}
disam.seuratmerge$condition <- ifelse(
  disam.seuratmerge$orig.ident %in% c("control1", "control2", "control3", "patient1", "patient2", "patient4"), "Control",
  ifelse(disam.seuratmerge$orig.ident %in% c("Covid_1", "Covid_2_sample1", "Covid_2_sample2", "Covid_4", "Covid_5", "Covid_6", "Presbyosmic_1", "Presbyosmic_3"), "Patient", "Other"))

disam.seuratmerge$study <- ifelse(
  disam.seuratmerge$orig.ident %in% c("Covid_1", "Covid_2_sample1", "Covid_2_sample2", "Covid_4", "Covid_5", "Covid_6"), "Covid",
  ifelse(disam.seuratmerge$orig.ident %in% c("patient1", "patient2", "patient4"), "Patients",
         ifelse(disam.seuratmerge$orig.ident %in% c("Presbyosmic_1", "Presbyosmic_3", "control1", "control2", "control3"), "PreControl", "Other"))
)

disam.seuratmerge$Disease_type <-  ifelse(
  disam.seuratmerge$orig.ident %in% c("Covid_1", "Covid_2_sample1", "Covid_2_sample2", "Covid_4", "Covid_5", "Covid_6"), "Covid",
  ifelse(disam.seuratmerge$orig.ident %in% c("Presbyosmic_1", "Presbyosmic_3"),"Presbyosmic", "Healthy"))

disam.seuratmerge$FH_group <- ifelse(
  disam.seuratmerge$orig.ident %in% c("control1", "control2", "control3"),"Control",
      ifelse(disam.seuratmerge$orig.ident %in% c("Presbyosmic_1", "Presbyosmic_3", "Covid_1", "Covid_2_sample1", "Covid_2_sample2", "Covid_4", "Covid_5", "Covid_6"),"Disease", "patients"))
```

```{r}
disam.seuratmerge$orig.ident <- factor(disam.seuratmerge$orig.ident, c("control1", "control2", "control3", "patient1", "patient2", "patient4", "Covid_1", "Covid_2_sample1", "Covid_2_sample2", "Covid_4", "Covid_5", "Covid_6", "Presbyosmic_1", "Presbyosmic_3"))

DimPlot(disam.seuratmerge, reduction = 'harmony_umap', split.by = 'orig.ident', ncol = 4, order = 'condition', group.by = 'label')
```

```{r}
avgexp_iosn <- AverageExpression(subset(disam.seuratmerge, label == "iOSN"), assays = "SCT", group.by = "orig.ident")$SCT
pheatmap::pheatmap(cor(as.matrix(avgexp_iosn)), cluster_rows = T, cluster_cols = T)

avgexp_mosn <- AverageExpression(subset(disam.seuratmerge, label == "mOSN"), assays = "SCT", group.by = "orig.ident")$SCT
pheatmap::pheatmap(cor(as.matrix(avgexp_mosn)), cluster_rows = T, cluster_cols = T)

avgexp <- AverageExpression(disam.seuratmerge, assays = "SCT", group.by = "orig.ident")$SCT
pheatmap::pheatmap(cor(as.matrix(avgexp)), cluster_rows = T, cluster_cols = T)

new_labels <- c(
  "control1", "control2", "control3",
  "control4", "control5", "control6",
  "Covid_1", "Covid_2_sample1", "Covid_2_sample2", "Covid_4", "Covid_5", "Covid_6",
  "Presbyosmic_1", "Presbyosmic_3"
)

cor_matrix <- cor(as.matrix(avgexp))

rownames(cor_matrix) <- new_labels
colnames(cor_matrix) <- new_labels
pheatmap::pheatmap(
  cor_matrix,
  cluster_rows = TRUE,
  cluster_cols = TRUE
)
```

##Differential Expression
##Compare with removed patient1,4, Covid_2_sample1 and Covid_2_sample2
```{r}
Idents(disam.seuratmerge) <- "condition"
to.keep <- disam.seuratmerge@meta.data %>%
  filter(!orig.ident %in% c("patient1","patient4", "Covid_2_sample1", "Covid_2_sample2")) %>%
  rownames()

diff_merge <- PrepSCTFindMarkers(subset(disam.seuratmerge, cells = to.keep))

options(future.globals.maxSize = 8 * 1024^3)
diff_exp <- FindMarkers(diff_merge, ident.1 = "Patient", ident.2 = "Control", verbose = FALSE, assay = "SCT", logfc.threshold = 0.25, min.pct = 0.1)

diff_exp$gene <- rownames(diff_exp)
diff_exp$direction <- ifelse(
  diff_exp$p_val_adj < 0.05 & diff_exp$avg_log2FC > 0.25, "Upregulated",
  ifelse(diff_exp$p_val_adj < 0.05 & diff_exp$avg_log2FC < -0.25, "Downregulated", "Not Significant"))

table(diff_exp$direction)
```

##Check for AIS genes in up or down regulated genes
```{r}
Upregulated <- diff_exp %>% filter(direction == 'Upregulated') %>% pull(gene)
Downregulated <- diff_exp %>% filter(direction == 'Downregulated') %>% pull(gene)

AIS <- read.csv('../originals/Matt_AIS_list.csv')
AIS <- unique(AIS)
AIS$human_gene <- NA
out <- gprofiler2::gorth(AIS$Genes,
        source_organism = "mmusculus",
        target_organism = "hsapiens")
AIS$human_gene[out$input_number] <- out$ortholog_name

AIS_downregulated <- intersect(AIS$human_gene, Downregulated)
AIS_upregulated <- intersect(AIS$human_gene, Upregulated)
```

```{r}
Idents(disam.seuratmerge) <- "condition"
VlnPlot(disam.seuratmerge, AIS_upregulated[1:6])
VlnPlot(disam.seuratmerge, AIS_upregulated[7:10])
```

##Volcano Plot
```{r}
##average expression
#avg_exp <- Matrix::rowMeans(disam.seuratmerge@assays$SCT@data)
#diff_exp$avg_expression <- avg_exp[rownames(diff_exp)]

top_genes <- diff_exp %>%
  filter(direction != "Not Significant") %>%
  arrange(p_val_adj) %>%
  head(10)

AIS_genes <- diff_exp %>%
  filter(rownames(diff_exp) %in% AIS_upregulated) %>%
  arrange(p_val_adj)

ggplot(diff_exp, aes(x = avg_log2FC, y = -log10(p_val_adj), colour = direction)) +
  geom_point(alpha = 0.8, size = 1.5) +
  scale_color_manual(values = c("Upregulated" = "red", 
                                "Downregulated" = "blue", 
                                "Not Significant" = "grey")) +
  geom_vline(xintercept = c(-0.25, 0.25), linetype = "dashed", colour = "black") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", colour = "black") +
  theme_minimal() +
  labs(title = "Patient vs Control",
       x = "Log2 Fold Change",
       y = "-Log10 Adjusted P-value",
       colour = "Significance") +
  theme(plot.title = element_text(hjust = 0.5)) + 
  geom_text_repel(
  data = AIS_genes,
  aes(x = avg_log2FC, y = -log10(p_val_adj), label = gene),
  size = 3,
  box.padding = 1.5,
  point.padding = 1.5,
  max.overlaps = Inf,
  force = 2,
  direction = "y",             
  hjust = 0,                   
  nudge_x = 1.5,
  segment.size = 0.2
)
```

```{r}
Idents(disam.seuratmerge) <- "condition"
VlnPlot(disam.seuratmerge, c("TUBB3"))

VlnPlot(
  disam.seuratmerge,
  features = "TUBB3",
  idents = setdiff(levels(disam.seuratmerge), c("patient1", "patient4", "control1"))
)

Idents(disam.seuratmerge) <- "orig.ident"
VlnPlot(disam.seuratmerge, c("TUBB3"))
```

##Gene Ontology
```{r}
universe <- rownames(disam.seuratmerge)

up_regulated <- diff_exp %>% filter(direction == 'Upregulated') %>% pull(gene)
down_regulated <- diff_exp %>% filter(direction == 'Downregulated') %>% pull(gene)

diff_exp_upgo <- enrichGO(gene = up_regulated,
                   ont = "BP",
                   keyType = "SYMBOL",
                   OrgDb = "org.Hs.eg.db",
                   universe = universe,
                   pvalueCutoff = 0.05,
                   qvalueCutoff = 0.05)

diff_exp_upgo_results <- diff_exp_upgo@result

diff_exp_downgo <- enrichGO(gene = down_regulated,
                   ont = "BP",
                   keyType = "SYMBOL",
                   OrgDb = "org.Hs.eg.db",
                   universe = universe,
                   pvalueCutoff = 0.05,
                   qvalueCutoff = 0.05)

diff_exp_downgo_results <- diff_exp_downgo@result

AIS_up <- enrichGO(gene = AIS_upregulated,
                   ont = "BP",
                   keyType = "SYMBOL",
                   OrgDb = "org.Hs.eg.db",
                   universe = universe,
                   pvalueCutoff = 0.05,
                   qvalueCutoff = 0.05)
AIS_up_results <- AIS_up@result %>% arrange(p.adjust)
```

```{r}
diff_exp_upgo_results %>% arrange(p.adjust) %>% arrange(pvalue) #%>% slice_head(n =10)
diff_exp_downgo_results %>% arrange(p.adjust) %>% arrange(pvalue)# %>% slice_head(n =10)
```

##RRVGO
```{r}
simMatrix_diffup <- calculateSimMatrix(diff_exp_upgo_results$ID,
                                orgdb="org.Hs.eg.db",
                                ont="BP",
                                method="Rel")

scores_diffup <- setNames(-log10(diff_exp_upgo_results$p.adjust), diff_exp_upgo_results$ID)

rvgo_diffup <- reduceSimMatrix(simMatrix_diffup, scores_diffup, threshold = 0.7, orgdb = "org.Hs.eg.db")
```

```{r}
rvgo_diffup_filtered <- rvgo_diffup %>% filter(go == parent)
```

```{r}
simMatrix_diffdown <- calculateSimMatrix(diff_exp_downgo_results$ID,
                                orgdb="org.Hs.eg.db",
                                ont="BP",
                                method="Rel")

scores_diffdown <- setNames(-log10(diff_exp_downgo_results$p.adjust), diff_exp_downgo_results$ID)

rvgo_diffdown <- reduceSimMatrix(simMatrix_diffdown, scores_diffdown, threshold = 0.7, orgdb = "org.Hs.eg.db")
```

```{r}
rvgo_diffdown_filtered <- rvgo_diffdown %>% filter(go == parent)
```

###AIS RVGO
```{r}
simMatrix_diffAIS <- calculateSimMatrix(AIS_up_results$ID,
                                orgdb="org.Hs.eg.db",
                                ont="BP",
                                method="Rel")

scores_diffAIS <- setNames(-log10(AIS_up_results$p.adjust), AIS_up_results$ID)

rvgo_diffAIS <- reduceSimMatrix(simMatrix_diffAIS, scores_diffAIS, threshold = 0.7, orgdb = "org.Hs.eg.db")

rvgo_diffAIS_filtered <- rvgo_diffAIS %>% filter(go == parent)
```

##Plot gene ontology
```{r}
plot_function3 <- function(df, title) {
  df %>% 
  slice_head(n =10) %>%
  ggplot(aes(x = score, y = reorder(term, score), fill = score)) +  
    geom_bar(stat = "identity", width = 0.5) +  
    scale_fill_gradient(low = "blue", high = "red") +  
    theme_minimal() +
    labs(x = "Score", y = "Term", fill = "Score", title = title) +
    theme(
      axis.text.y = element_text(size = 10, color = "black"),  # Bigger and darker terms
      axis.text.x = element_text(size = 12, color = "black"),
      axis.title = element_text(size = 14, color = "black"),
      plot.title = element_text(size = 16, face = "bold", color = "black")
    )
}

#diff_exp_downgo_results[c(1,2,3,5,6,7,8,12,13,14),]

plot_function3(rvgo_diffdown_filtered[c(1:6, 9, 13:15),], paste("Down-regulated Terms"))
plot_function3(rvgo_diffup_filtered, paste("Up-regulated Terms"))
plot_function3(rvgo_diffAIS_filtered, paste("AIS Up-regulated Terms"))
```

##Compare between diseases
```{r}
Idents(diff_merge) <- "Disease_type"
diff_exp_disease <- FindMarkers(diff_merge, ident.1 = "Covid", ident.2 = "Presbyosmic", verbose = FALSE, assay = "SCT", logfc.threshold = 0.25, min.pct = 0.1)

diff_exp_disease$gene <- rownames(diff_exp_disease)
diff_exp_disease$direction <- ifelse(
  diff_exp_disease$p_val_adj < 0.05 & diff_exp_disease$avg_log2FC > 0.25, "Upregulated",
  ifelse(diff_exp_disease$p_val_adj < 0.05 & diff_exp_disease$avg_log2FC < -0.25, "Downregulated", "Not Significant"))

Upregulated_patients <- diff_exp_disease %>% filter(direction == "Upregulated")
Downregulated_patients <- diff_exp_disease %>% filter(direction == "Downregulated")
intersect(AIS, Downregulated_patients)
intersect(AIS, Upregulated_patients)
```

##Compare covid
```{r}
Idents(diff_merge) <- "Disease_type"
diff_exp_covid <- FindMarkers(diff_merge, ident.1 = "Covid", ident.2 = "Healthy", verbose = FALSE, assay = "SCT", logfc.threshold = 0.25, min.pct = 0.1)

diff_exp_covid$gene <- rownames(diff_exp_covid)
diff_exp_covid$direction <- ifelse(
  diff_exp_covid$p_val_adj < 0.05 & diff_exp_covid$avg_log2FC > 0.25, "Upregulated",
  ifelse(diff_exp_covid$p_val_adj < 0.05 & diff_exp_covid$avg_log2FC < -0.25, "Downregulated", "Not Significant"))

table(diff_exp_covid$direction)

Upregulated_covid <- diff_exp_covid %>% filter(direction == 'Upregulated') %>% pull(gene)
Downregulated_covid <- diff_exp_covid %>% filter(direction == 'Downregulated') %>% pull(gene)

intersect(AIS$human_gene, Downregulated_covid)
AIS_upregulated_covid <- intersect(AIS$human_gene, Upregulated_covid)

AIS_genes_covid <- diff_exp_covid %>%
  filter(rownames(diff_exp_covid) %in% AIS_upregulated_covid) %>%
  arrange(p_val_adj)

covid_vlc <- ggplot(diff_exp_covid, aes(x = avg_log2FC, y = -log10(p_val_adj), colour = direction)) +
  geom_point(alpha = 0.8, size = 1.5) +
  scale_color_manual(values = c("Upregulated" = "red", 
                                "Downregulated" = "blue", 
                                "Not Significant" = "grey")) +
  geom_vline(xintercept = c(-0.25, 0.25), linetype = "dashed", colour = "black") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", colour = "black") +
  theme_minimal() +
  labs(title = "COVID vs Control",
       x = "Log2 Fold Change",
       y = "-Log10 Adjusted P-value",
       colour = "Significance") +
  theme(plot.title = element_text(hjust = 0.5)) + 
  geom_text_repel(
  data = AIS_genes_covid,
  aes(x = avg_log2FC, y = -log10(p_val_adj), label = gene),
  size = 3,
  box.padding = 1,
  point.padding = 0.5,
  max.overlaps = Inf,
  direction = "y",             
  hjust = 0,                   
  nudge_x = 1.5                
)
```

##Compare presbyosmic
```{r}
Idents(diff_merge) <- "Disease_type"
diff_exp_presbyosmic <- FindMarkers(diff_merge, ident.1 = "Presbyosmic", ident.2 = "Healthy", verbose = FALSE, assay = "SCT", logfc.threshold = 0.25, min.pct = 0.1)

diff_exp_presbyosmic$gene <- rownames(diff_exp_presbyosmic)
diff_exp_presbyosmic$direction <- ifelse(
  diff_exp_presbyosmic$p_val_adj < 0.05 & diff_exp_presbyosmic$avg_log2FC > 0.25, "Upregulated",
  ifelse(diff_exp_presbyosmic$p_val_adj < 0.05 & diff_exp_presbyosmic$avg_log2FC < -0.25, "Downregulated", "Not Significant"))

table(diff_exp_presbyosmic$direction)

Upregulated_presbyosmic <- diff_exp_presbyosmic %>% filter(direction == 'Upregulated') %>% pull(gene)
Downregulated_presbyosmic <- diff_exp_presbyosmic %>% filter(direction == 'Downregulated') %>% pull(gene)

intersect(AIS$human_gene, Downregulated_presbyosmic)
intersect(AIS$human_gene, Upregulated_presbyosmic)

AIS_genes_presbyosmic <- diff_exp_presbyosmic %>%
  filter(rownames(diff_exp_presbyosmic) %in% intersect(AIS$human_gene, Upregulated_presbyosmic)) %>%
  arrange(p_val_adj)

pre_vlc <- ggplot(diff_exp_presbyosmic, aes(x = avg_log2FC, y = -log10(p_val_adj), colour = direction)) +
  geom_point(alpha = 0.8, size = 1.5) +
  scale_color_manual(values = c("Upregulated" = "red", 
                                "Downregulated" = "blue", 
                                "Not Significant" = "grey")) +
  geom_vline(xintercept = c(-0.25, 0.25), linetype = "dashed", colour = "black") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", colour = "black") +
  theme_minimal() +
  labs(title = "Presbyosmic vs Control",
       x = "Log2 Fold Change",
       y = "-Log10 Adjusted P-value",
       colour = "Significance") +
  theme(plot.title = element_text(hjust = 0.5)) + 
  geom_text_repel(
  data = AIS_genes_presbyosmic,
  aes(x = avg_log2FC, y = -log10(p_val_adj), label = gene),
  size = 3,
  box.padding = 1,
  point.padding = 0.5,
  max.overlaps = Inf,
  direction = "y",             
  hjust = 0,                   
  nudge_x = 1.5                
)
```

##Compare between the two
```{r}
covid <- diff_exp_covid %>% filter(direction != "Not Significant")
presbyosmic <- diff_exp_presbyosmic %>% filter(direction != "Not Significant")

setdiff(rownames(covid), rownames(presbyosmic))
setdiff(rownames(presbyosmic), rownames(covid))

intersect(AIS, setdiff(rownames(presbyosmic), rownames(covid))) #No AIS genes present
intersect(rownames(presbyosmic), rownames(covid))
```

#Calculate proportions of cells
```{r}
disease_all.list <- readRDS("../outputs/RDS_files/disease/disease.seuratlistMS2.rds")
samples_all.list <- readRDS("../outputs/RDS_files/samples/samples.seuratlistMS.rds")
disease.subset_OSN <- readRDS("../outputs/RDS_files/disease/disease.subset_OSN.rds")
samples.seurat <- readRDS('../outputs/RDS_files/samples/samples.seuratpseudorank_june.rds')
```

##check OSN cell identities in samples and calculate proportions
```{r}
samples.seurat_list <- SplitObject(samples.seurat, split.by = 'orig.ident')

#remove objects that are no longer part of analysis
samples_all.list[[6]] <- NULL
```

```{r}
OSN_list_samples <- lapply(samples.seurat_list, function(x) {
  keep <- length(colnames(x))
  return(keep)
})

allCT_list_samples <- lapply(samples_all.list, function(x) {
  keep <- length(colnames(x))
  return(keep)
})

props_samples <- mapply(function(x, y) {
  (x / y) * 100
}, OSN_list_samples, allCT_list_samples)
```

##check OSN cell identities in disease and calculate proportions
```{r}
disease_subset_list <- SplitObject(disease.subset_OSN, split.by = 'orig.ident')

#merge back sample 2
disease_subset_list[["Covid_2"]] <- merge(x = disease_subset_list[[7]], y = disease_subset_list[[8]], merge.data = TRUE)

#remove remaining sample after merge
disease_subset_list[["Covid_2_sample1"]] <- NULL
disease_subset_list[["Covid_2_sample2"]] <- NULL

#remove objects that are no longer part of analysis
disease_all.list[["Covid_3"]] <- NULL
disease_all.list[["Presbyosmic_2"]] <- NULL

```

```{r}
OSN_list_disease <- lapply(disease_subset_list, function(x) {
  keep <- length(colnames(x))
  return(keep)
})

allCT_list_disease <- lapply(disease_all.list, function(x) {
  keep <- length(colnames(x))
  return(keep)
})

props_disease <- mapply(function(x, y) {
  (x / y) * 100
}, OSN_list_disease, allCT_list_disease)
```

##check difference
```{r}
avg_prop_samples <- mean(props_samples)
avg_prop_disease <- mean(props_disease)

wilcox.test(props_samples, props_disease)
```

```{r}
OSN_proportions <- c(props_samples, props_disease)

OSN_proportions <- data.frame(
  Sample = factor(names(OSN_proportions), levels = names(OSN_proportions)),
  Proportion = as.numeric(OSN_proportions)
)

OSN_proportions$Condition <- ifelse(OSN_proportions$Sample %in% c("control1", "control2", "control3", "patient1", "patient2", "patient4"), "Control",
                             ifelse(OSN_proportions$Sample %in% c("Covid_1", "Covid_2", "Covid_4", "Covid_5", "Covid_6"), "Covid",
                             ifelse(OSN_proportions$Sample %in% c("Presbyosmic_1", "Presbyosmic_3"), "Presbyosmic", "Other")))

ggplot(OSN_proportions, aes(x = Condition, y = Proportion, fill = Proportion)) +
  geom_col(width = 0.5, color = "black") +
  scale_fill_gradient(
    low = "lightblue",  # light muted teal
    high = "#8C6BB1"  # muted purple
  ) +
  labs(
    title = "Proportion of OSNs",
    x = "Sample",
    y = "Proportion"
  ) +
  theme_minimal(base_size = 14) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
    plot.title = element_text(face = "bold", hjust = 0.5),
    panel.grid.major.x = element_blank(),
    legend.position = "none"
  )
```

```{r}
OSN_proportions$Sample <- sub("patient1", "control4", OSN_proportions$Sample)
OSN_proportions$Sample <- sub("patient2", "control5", OSN_proportions$Sample)
OSN_proportions$Sample <- sub("patient4", "control6", OSN_proportions$Sample)

prop_bar <- ggplot(OSN_proportions, aes(x = factor(Sample, levels = c("control1", "control2", "control3", "control4", "control5", "control6", "Covid_1", "Covid_2", "Covid_4", "Covid_5", "Covid_6", "Presbyosmic_1", "Presbyosmic_3")), y = Proportion, fill = Condition)) +
  geom_col(width = 0.5, color = "black") +
  scale_fill_brewer(palette = "Set2") +
  labs(
    title = "Proportion of OSNs",
    x = "Sample",
    y = "Proportion %"
  ) +
  theme_minimal(base_size = 14) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
    plot.title = element_text(face = "bold", hjust = 0.5),
    panel.grid.major.x = element_blank()
  )

```

```{r}
Idents(disam.seuratmerge) <- "orig.ident"
VlnPlot(disam.seuratmerge, features = AIS_upregulated)
```