---
title: "Pseudotime_Analysis"
author: "Srilakshmi"
date: "2025-09-01"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Pseudo time
##Create DiffMap
```{r}
data <- as.matrix(samples.seurat@reductions$harmony_int@cell.embeddings)

dm <- DiffusionMap(data, n_pcs = 3, n_eigs = 2)
```

```{r}
samples.seurat[["diffmap"]] <- CreateDimReducObject(embeddings = dm@eigenvectors[,1:2], key = "dc_", assay = "SCT")

DimPlot(samples.seurat, reduction = "diffmap", group.by = 'label') + labs(title = "DiffusionMap")

```

##Apply slignshot
```{r}
sds <- slingshot(Embeddings(samples.seurat, "diffmap"), 
                 clusterLabels = samples.seurat$seurat_clusters,
                 start.clus= '4')
slingLineages(sds)
```

##Calculate average pseudotime
```{r}
sds <- readRDS("../outputs/RDS_files/pseudotime/sds_human_june.rds")
samples.seurat$pseudotime <- slingAvgPseudotime(sds)

FeaturePlot(samples.seurat, "pseudotime", reduction = "harmony_int_umap", label = TRUE)

FeatureScatter(samples.seurat, feature1 = "pseudotime", feature2 = "STMN1") + geom_smooth()
FeatureScatter(samples.seurat, feature1 = "pseudotime", feature2 = "ANK3") + geom_smooth()
```

## Calculate proportions of cells 
```{r}
meta <- samples.seurat@meta.data
meta$bins <- cut(samples.seurat$pseudotime, breaks = 10, include.lowest = TRUE)

meta$bin_label <- factor(meta$bins, labels = paste0("Bin ", seq_len(nlevels(meta$bins))))

prop <- meta %>% 
  group_by(label, bin_label) %>%
  summarise(n = n(), .groups = "drop") %>%
  group_by(bin_label) %>% 
  mutate(proportion = (n / sum(n))*100)
```

```{r}
ggplot(prop, aes(x = bin_label, y = proportion, colour = label)) +
  geom_point(size = 3) +
  theme_minimal() +
  labs(x = "Pseudotime Bins", y = "Proportion of cells", colour = "Label")
```

##Pseudo-ranking
```{r}
samples.seurat$pseudorank <- (rank(samples.seurat$pseudotime)*10)/696

metarank <- samples.seurat@meta.data
metarank$bins <- cut(samples.seurat$pseudorank, breaks = 10, include.lowest = TRUE)

metarank$bin_label <- factor(metarank$bins, labels = paste0("Bin ", seq_len(nlevels(meta$bins))))

prop_rank <- metarank %>% 
  group_by(label, bin_label) %>%
  summarise(n = n(), .groups = "drop") %>%
  group_by(bin_label) %>% 
  mutate(proportion = (n / sum(n))*100)

ggplot(prop_rank, aes(x = bin_label, y = proportion, colour = label)) +
  geom_point(size = 3) +
  theme_minimal() +
  labs(x = "Pseudo-rank Bins", y = "Proportion of cells", colour = "Label")
```

##Plot densities
```{r}
pseudotime_density <- ggplot(meta, aes(x = pseudotime, fill = label, colour = label)) +
  geom_density(alpha = 0.4) +
  theme_minimal() +
  labs(x = "Pseudotime", y = "Density", fill = "Label", colour = "Label")

pseudorank_density <- ggplot(metarank, aes(x = pseudorank, fill = label, colour = label)) +
  geom_density(alpha = 0.4) +
  theme_minimal() +
  labs(x = "Pseudorank", y = "Density", fill = "Label", colour = "Label")
```

# tradeSeq
```{r}
## identify a distinct type of de pattern along a lineage
weights <- slingCurveWeights(sds)
weights <- rowMaxs(weights)
counts <- as.matrix(samples.seurat@assays$SCT$counts)
```

##fitGAM
```{r}
#analyse relationship between gene expression and pseudotime using fitGAM
sce <- fitGAM(counts = counts, 
              pseudotime = samples.seurat[["pseudorank"]], 
              cellWeights = weights)
```

## startVsEndTest
```{r}
sce <- readRDS("../outputs/RDS_files/pseudotime/sce_june.rds")
startRes <- startVsEndTest(sce)
```

###Check for AIS genes
```{r}
AIS_MG <- read.csv("../originals/Matt_AIS_list.csv")
AIS_MG <- unique(AIS_MG)

AIS_MG$human_gene <- NA
out <- gprofiler2::gorth(AIS_MG$Genes,
        source_organism = "mmusculus",
        target_organism = "hsapiens")
AIS_MG$human_gene[out$input_number] <- out$ortholog_name

res <- rownames(startRes)
check_MG <- unique(AIS_MG$human_gene)

AIS_MG_check <- intersect(res, check_MG)

startRes %>% filter(res %in% AIS_MG_check) %>% filter(pvalue < 0.05)
```

### Filter list
```{r}
sr_sd <- sd(startRes$waldStat)

quant <- quantile(startRes$waldStat, 0.975)

startRes_filtered <- startRes %>%
  filter(pvalue < 0.05) %>%
  filter(waldStat > quant) %>%
  arrange(desc(logFClineage1))

#Top 10 top differentially expressed genes
#startRes %>% 
#  filter(pvalue < 0.05) %>% 
#  filter(waldStat > quant) %>%
#  slice_max(logFClineage1, n = 10)
```

```{r}
Idents(samples.seurat) <- "label"
FeatureScatter(samples.seurat, "pseudorank", "JAKMIP1") +geom_smooth() 
FeatureScatter(samples.seurat, "pseudotime", "CD47") +geom_smooth()
```

```{r}
up <- startRes_filtered %>% filter(startRes_filtered$logFClineage1 > 0.5)

down <- startRes_filtered %>% filter(startRes_filtered$logFClineage1 < -0.5)
```

###Check how many are significant and are up and down regulated
```{r}
startRes_filtered %>% filter(rownames(startRes_filtered) %in% AIS_MG_check)

up %>% filter(rownames(up) %in% AIS_MG_check)

down %>% filter(rownames(down) %in% AIS_MG_check)
```

```{r}
Idents(samples.seurat) <- "label"
FeatureScatter(samples.seurat, "pseudorank", "FGF12") + geom_smooth()
FeatureScatter(samples.seurat, "pseudorank", "MAP7D2") + geom_smooth()
FeatureScatter(samples.seurat, "pseudorank", "MAPRE3") + geom_smooth()
FeatureScatter(samples.seurat, "pseudorank", "EPB41") + geom_smooth()

FeatureScatter(samples.seurat, "pseudorank", "SCN9A") + geom_smooth()
FeatureScatter(samples.seurat, "pseudorank", "SCN3B") + geom_smooth()
```

###Gene Ontology
```{r}
library(clusterProfiler)
#BiocManager::install("org.Hs.eg.db")

universe <- rownames(samples.seurat)

up_go <- enrichGO(gene = rownames(up),
                   ont = "BP",
                   keyType = "SYMBOL",
                   OrgDb = "org.Hs.eg.db",
                   universe = universe,
                   pvalueCutoff = 0.05,
                   qvalueCutoff = 0.05)

down_go <- enrichGO(gene = rownames(down),
                   ont = "BP",
                   keyType = "SYMBOL",
                   OrgDb = "org.Hs.eg.db",
                   universe = universe,
                   pvalueCutoff = 0.05,
                   qvalueCutoff = 0.05)
```

###RVGO
```{r}
up_results <- up_go@result

simMatrix <- calculateSimMatrix(up_results$ID,
                                orgdb="org.Hs.eg.db",
                                ont="BP",
                                method="Rel")

scores <- setNames(-log10(up_results$p.adjust), up_results$ID)

rvgo_up <- reduceSimMatrix(simMatrix, scores, threshold = 0.7, orgdb = "org.Hs.eg.db")
```

```{r}
rvgo_up_filtered <- rvgo_up <- rvgo_up %>% filter(go == parent)
```

```{r}
down_results <- down_go@result
simMatrix_down <- calculateSimMatrix(down_results$ID,
                                orgdb="org.Hs.eg.db",
                                ont="BP",
                                method="Rel")

scores_down <- setNames(-log10(down_results$p.adjust), down_results$ID)

rvgo_down <- reduceSimMatrix(simMatrix_down, scores_down, threshold = 0.7, orgdb = "org.Hs.eg.db")
```

```{r}
rvgo_down_filtered <- rvgo_down %>% filter(go == parent)
```

```{r}
plot_function <- function(df, title) {
  ggplot(df, aes(x = score, y = reorder(term, score), fill = score)) +  
    geom_bar(stat = "identity", width = 0.5) +  
    scale_fill_gradient(low = "blue", high = "red") +  
    theme_minimal() +
    labs(x = "Score", y = "Term", title = title, fill = "Score") +
    theme(
      axis.text.y = element_text(size = 12, color = "black"),  # Bigger and darker terms
      axis.text.x = element_text(size = 12, color = "black"),
      axis.title = element_text(size = 14, color = "black"),
      plot.title = element_text(size = 16, face = "bold", color = "black")
    )
}

plot_function_log <- function(df, title) {
  ggplot(df, aes(x = log10(score), y = reorder(term, score), fill = score)) +  
    geom_bar(stat = "identity") +  
    scale_fill_gradient(low = "blue", high = "red") +
    theme_minimal() +
    labs(x = "log10(Score)", y = "Term", title = title, fill = "Score") +
    theme(
      axis.text.y = element_text(size = 12, color = "black"),  # Bigger and darker terms
      axis.text.x = element_text(size = 12, color = "black"),
      axis.title = element_text(size = 14, color = "black"),
      plot.title = element_text(size = 16, face = "bold", color = "black")
    )
}

plot_function(rvgo_up_filtered[c(1:8, 10, 12),], paste("Enriched Cluster Up Regulated"))
plot_function_log(rvgo_down_filtered[c(1:9, 12),],paste("Enriched Cluster Down Regulated"))
```

##Associations test
```{r}
associationRes <- associationTest(sce)
associationRes <- na.omit(associationRes)

sd <- sd(associationRes$waldStat)
quant_2 <- quantile(associationRes$waldStat, 0.925)

association_filtered <- associationRes %>% 
  filter(pvalue < 0.05) %>%
  filter(waldStat > quant_2) %>%
  arrange(desc(meanLogFC))
  #slice_min(meanLogFC, n = 10)
```

```{r}
up_association <- rownames(association_filtered)[association_filtered$meanLogFC > 0.5]

startRes_fold <- startRes %>%
  filter(logFClineage1 < 0.2) %>%
  filter(logFClineage1 > -0.2)

intersect <- intersect(up_association, rownames(startRes_fold))

down_association <- rownames(association_filtered)[association_filtered$meanLogFC < -0.5]
```

```{r}
association_filtered[rownames(association_filtered) %in% intersect,]
Idents(samples.seurat) <- "label"
FeatureScatter(samples.seurat, "pseudorank", "BPIFB1") + geom_smooth()
FeatureScatter(samples.seurat, "pseudorank", "CNTRL") + geom_smooth()
FeatureScatter(samples.seurat, "pseudorank", "KTN1") + geom_smooth()
FeatureScatter(samples.seurat, "pseudorank", "RDX") + geom_smooth()
FeatureScatter(samples.seurat, "pseudorank", "AC092667.1") + geom_smooth()
```

## tradeseq with mice
```{r}
mice_sample <- readRDS("../outputs/RDS_files/pseudotime/obj_harmony.rds")
mice_sce <- readRDS("../outputs/RDS_files/pseudotime/sce_mice.rds")
```

```{r}
data_mice <- as.matrix(mice_sample@reductions$harmony@cell.embeddings)

dm <- DiffusionMap(data_mice, n_pcs = 3)
```

```{r}
mice_sample[["diffmap"]] <- CreateDimReducObject(embeddings = dm@eigenvectors[,1:3], key = "dc_", assay = "SCT")

DimPlot(mice_sample, reduction = "diffmap", label = TRUE) 
```

```{r}
sds_mice <- slingshot(Embeddings(mice_sample, "diffmap"), 
                 clusterLabels = mice_sample$seurat_clusters,
                 start.clus= '13')
slingLineages(sds_mice)
```

```{r}
#add pseudotime for all 4 lineages
mice_sample@meta.data <- cbind(mice_sample@meta.data, slingPseudotime(sds_mice, na = FALSE))
#add averaged
mice_sample$pseudotime <- slingAvgPseudotime(sds_mice)

FeaturePlot(mice_sample, "pseudotime", reduction = "harmony", label = TRUE)
```

##startvsend test with averaged pseudotime
```{r}
startRes_mice <- startVsEndTest(mice_sce)

quant2 <- quantile(startRes_mice$waldStat, 0.975)

startRes_filtered_mice <- startRes_mice %>%
  filter(pvalue < 0.05) %>%
  filter(waldStat > quant2) %>%
  arrange(desc(logFClineage1))

up_mice <- rownames(startRes_filtered_mice)[startRes_filtered_mice$logFClineage1 > 0.5]

down_mice <- rownames(startRes_filtered_mice)[startRes_filtered_mice$logFClineage1 < -0.5]

mice_sve <- c(as.data.frame(up_mice), as.data.frame(down_mice))

startRes_filtered_mice %>%
  filter(logFClineage1 < 0.5) %>%
  filter(logFClineage1 > -0.5)

AIS_MG <- read.csv("../originals/Matt_AIS_list.csv")
AIS_MG <- unique(AIS_MG$Genes)

res <- rownames(startRes_filtered_mice)
check_MG <- unique(AIS_MG)

AIS_MG_check <- intersect(res, check_MG)

startRes_filtered_mice %>% filter(res %in% AIS_MG_check) %>% filter(pvalue < 0.05)
```

###Compare genes in mice and human
```{r}
up_mice <- as.data.frame(up_mice)
colnames(up_mice)[colnames(up_mice) == "up_mice"] <- "gene"

up_convert <- gprofiler2::gorth(up_mice$gene,
                         source_organism = "mmusculus",
                         target_organism = "hsapiens")

down_mice <- as.data.frame(down_mice)
colnames(down_mice)[colnames(down_mice) == "down_mice"] <- "gene"

down_convert <- gprofiler2::gorth(down_mice$gene,
                         source_organism = "mmusculus",
                         target_organism = "hsapiens")

up_common <- intersect(up, up_convert$ortholog_name)
down_common <- intersect(down, down_convert$ortholog_name)

up_unique <- setdiff(up, up_convert$ortholog_name)
up_convert_unique <- setdiff(up_convert$ortholog_name, up)
```

##Association test with mice
```{r}
associationRes_mice <- associationTest(mice_sce)
associationRes_mice <- na.omit(associationRes_mice)

sd <- sd(associationRes_mice$waldStat)
quant_2 <- quantile(associationRes_mice$waldStat, 0.925)

association_filtered_mice <- associationRes_mice %>% 
  filter(pvalue < 0.05) %>%
  filter(waldStat > quant_2) %>%
  arrange(desc(meanLogFC))
  #slice_min(meanLogFC, n = 10)
```

```{r}
up_2 <- rownames(association_filtered_mice)[association_filtered_mice$meanLogFC > 0.5]

startRes_fold <- startRes_mice %>%
  filter(logFClineage1 < 0.2) %>%
  filter(logFClineage1 > -0.2)

intersect <- intersect(up_2, rownames(startRes_fold))

down_2 <- rownames(association_filtered_mice)[association_filtered_mice$meanLogFC < -0.5]
```

```{r}
asso_intersect <- association_filtered_mice[rownames(association_filtered_mice) %in% intersect,]

res <- rownames(asso_intersect)
check_MG <- unique(AIS_MG)

AIS_MG_check <- intersect(res, check_MG)

asso_intersect %>% filter(res %in% AIS_MG_check) %>% filter(pvalue < 0.05)
```

```{r}
FeatureScatter(mice_sample, "pseudotime", "Gng3") + geom_smooth()
FeatureScatter(mice_sample, "pseudotime", "Casp6") + geom_smooth()
FeatureScatter(mice_sample, "pseudotime", "Eif1") + geom_smooth()
```

##subset for 5000 cells
```{r}
set.seed(123)
cells_to_keep <- sample(colnames(mice_sample), 5000)

mice_sample_subset <- subset(mice_sample, cells = cells_to_keep)

#sds_subset <- slingshot(Embeddings(mice_sample_subset, "diffmap"), 
                 #clusterLabels = mice_sample_subset$seurat_clusters,
                 #start.clus= '13')
#slingLineages(sds_subset)
```

```{r}
## identify a distinct type of de pattern along a lineage
weights <- slingCurveWeights(sds_mice)
#weights <- rowMaxs(weights)
cellWeights <- weights[cells_to_keep, , drop = FALSE]
counts <- as.matrix(mice_sample_subset@assays$SCT$counts)

#saveRDS(cellWeights, "../Rstudio/cellWeights.rds")
#saveRDS(counts, "../Rstudio/counts.rds")
#saveRDS(sds_mice, "../Rstudio/sds_mice.rds")
#saveRDS(mice_sample_subset, "../Rstudio/mice_sample_subset.rds")
```

```{r}
#analyse relationship between gene expression and pseudotime using fitGAM
sce_mice_subset <- fitGAM(counts = counts, 
              pseudotime = slingPseudotime(sds_mice, na = FALSE)[colnames(mice_sample_subset),],
              cellWeights = cellWeights)
```

```{r}
sce_mice_subset <- readRDS("../outputs/RDS_files/pseudotime/sce_mice_subset.rds")
counts <- readRDS("../outputs/RDS_files/pseudotime/counts.rds")
```

##diffend test for lineage comparisons
```{r}
endRes <- diffEndTest(sce_mice_subset, pairwise = TRUE)
```

### Filter for significance in entire list 
```{r}
endRes <- na.omit(endRes)
quantile_end <- quantile(endRes$waldStat, 0.975)
quantile(endRes$waldStat_1vs2, 0.975)

endRes_filtered <- endRes %>%
  filter(pvalue < 0.05) %>%
  filter(waldStat > quantile_end) %>%
  arrange(desc(waldStat))

intersect(AIS_MG, toupper(rownames(endRes_filtered)))
```

```{r}
plotSmoothers(sce_mice_subset, counts, "Ubb")
```